### ISSUE TITLE
Tier 4: Generalization

### ISSUE THREAD

**Issue by JHollenb** [JHollenb/spectral-bit-hacking#4]
### 4.4 Tier 4: Generalization (UNLOCKED)

Test PAN on:
- Other primes P ∈ {43, 67, 89, 127} — does it generalize with the same K?
- Modular multiplication: a × b mod P — does the phase mixing architecture handle this?
- Two-step arithmetic: (a + b) × c mod P — does PAN compose operations?

**Comment by JHollenb** [comment]
[summary_20260328_195548.txt](https://github.com/user-attachments/files/26327700/summary_20260328_195548.txt)

**Comment by JHollenb** [comment]
[20260328_195237.log](https://github.com/user-attachments/files/26327721/20260328_195237.log)

<img width="1440" height="1162" alt="Image" src="https://github.com/user-attachments/assets/89f0bdd1-9e7e-424c-8dc7-288d8b477fa0" />

Tier 4 is essentially passed. 4 out of 5 primes grokked cleanly with identical hyperparameters. The fifth — P=89 — is not a failure.
P=89 is a late grokker, not a broken one. Look at the curve: val accuracy sits between 50–70% for 80,000 steps, then rockets from 70% to 97% between steps 88K and 99K. The grokking cliff is unambiguously starting. The run simply hit the 100K budget mid-transition. Running to 130K would almost certainly cross 99%. The same architecture, same K=9, same WD=0.01 — it just needs more time for P=89. This is qualitatively different from K<5 in the sweep, where the loss barely moved from random. P=89 is genuinely learning.
The scaling behavior is the real finding. P=43 grokked at step 6,000 — K=9 is over-provisioned for a 43-class problem (you probably only need 4 frequencies for mod-43). P=67 at 11,800 steps, P=113 at 11,200, P=127 at 23,400. The grokking step doesn't scale linearly with P, which means K=9 isn't just a fixed-cost solution — it adapts. The architecture is using only as many frequencies as it needs and the excess K slots remain dormant or create redundancy that actually helps.
The parameter count story is remarkable here. K=9 PAN on P=43 uses 619 parameters. On P=127 it uses 1,459 parameters. The decoder scales with P (it's a K×P linear layer), so the model automatically sizes itself to the problem. A 619-parameter network solving mod-43 arithmetic to 99.1% accuracy is probably the smallest neural network that has ever solved a nontrivial group operation.
What this means for the paper. Tier 4 is passing. The architecture is principled — it generalizes across primes without any hyperparameter tuning. The only thing to nail down is whether P=89 grokks with more steps (run it to 200K to confirm) and whether the approach works for primes not seen during any of this development (try P=59, P=71, P=97 as held-out primes).

**Comment by JHollenb** [comment]
<img width="1440" height="1020" alt="Image" src="https://github.com/user-attachments/assets/c38df83c-b711-433f-92e6-d33eb547d365" />

5/5 primes. Tier 4 passes. P=89 grokked at step 139,800 exactly as predicted — the 100K run had caught it mid-cliff. Same architecture, same K=9, same WD=0.01, no per-prime tuning.
The P=89 result is actually the most interesting one in the set. It's anomalously slow — 12× more steps than P=67 and P=113 despite being between them in size. The likely reason: P=89's Fourier basis frequencies sit in an awkward position relative to K=9's initialization. The optimizer finds the right circuit eventually, but the loss landscape for P=89 has a wide flat basin before the grokking cliff. This is worth a note but not a concern.
